apps/els_lsp/src/els_refactorerl_utils.erl

```diff
@@ -0,0 +1,284 @@
+-spec make_diagnostics(any(), string()) -> [els_diagnostics:diagnostic()].
+make_diagnostics([{{_Path, From, To}, Name} | Tail], DiagMsg) ->
+  Range = #{ from => From, to => To },
+  Id = refactorerl_poi,
+  #{ data := PoiData, range := PoiRange} =
+                        els_poi:new(Range, application, Id, Name), 
+  RangeLS = els_protocol:range(PoiRange),
+  Message = list_to_binary(DiagMsg ++ " " ++ PoiData),
+  Severity = ?DIAGNOSTIC_WARNING,
+  Source = source_name(),
+  Diag = els_diagnostics:make_diagnostic(RangeLS, Message, Severity, Source), 
+  [ Diag | make_diagnostics(Tail, DiagMsg) ];
+
+make_diagnostics([], _) ->
+  [];
+
+make_diagnostics({ok, {result, [{result,[{group_by, {nopos, _}, list, L}]}]}}, 
+                                                                    DiagMsg) ->
+  make_diagnostics(L, DiagMsg).
+
+%%==============================================================================
+%% Internal Functions
+%%==============================================================================
+
+-spec disable_node(atom()) -> atom().
+%%@doc
+%% Disables the node given in paramter, also notifes the user.
+disable_node(Node) ->
+  Msg = "RefactorErl is disconnected!
+  Reload ELS after you fixed theRefactorErl node!",
+  notification(Msg, ?MESSAGE_TYPE_ERROR),
+
+  els_config:set(refactorerl, #{"node" => {Node, disabled}}),
+  {error, disabled}.
+
+%%@doc
+%% Calls the RefactorErl node, to check if it's alive using ri:ls()
+-spec check_node(atom()) -> {error, timeout}
+                          | {error, badrpc}
+                          | {error, other}
+                          | atom().
+check_node(Node) ->
+  Response = rpc:call(Node, ri, ls, [], 10000),
+  case Response of
+    {{ok, _}, {error, _}} ->
+      Node;
+    ok ->
+      Node;
+    {badrpc, timeout} ->
+      {error, timeout};
+    {badrpc, _} ->
+      {error, badrpc};
+    _ ->
+      {error, other}
+  end.
+
+%%@doc
+%% Tries to connect to a node. 
+%% When it status is validate, the node hasn't been checked yet, 
+%% so it will reports the success, and failure as well,
+%% 
+%% when retry, it won't report.
+-spec connect_node({validate | retry, atom()}) -> {error, disconnected}
+                                                     | atom().
+connect_node({Status, Node}) ->
+  case {Status, check_node(Node)} of
+    {validate, {error, _}} ->
+      notification("RefactorErl is not connected!", ?MESSAGE_TYPE_INFO),
+      els_config:set(refactorerl, #{"node" => {Node, disconnected}}),
+      {error, disconnected};
+    {retry, {error, _}} ->
+      els_config:set(refactorerl, #{"node" => {Node, disconnected}}),
+      {error, disconnected};
+    {_, Node} ->
+      notification("RefactorErl is connected!", ?MESSAGE_TYPE_INFO),
+      els_config:set(refactorerl, #{"node" => {Node, validated}}),
+      {ok, Node}
+  end.
+
+%%@doc
+%% Gets a request id from the RefactorErl node, and returns it 
+-spec request_id() -> nodedown | {reqid | string()}.
```

Comment:
Nit: For consistency I would probably have all these functions returning either an `ok` or an `error` tuple`. For example, here it would be `{ok, RequestId}` or `{error, Reason}`. I am not sure about the current signature, is the atom `reqid` occasionally returned instead of the string?

Reply:
No, it either returns `'nodown'` or `{reqid | string()}`. If I wrap out the request ID string out from the tuple, the place where I call it, than I need to re-wrap it, as RefactorErl needs it in this tupled form.

A suggestion of mine is to lets go with: `{error, Reason}` or `{ok, {reqid, RequestID}}` or stay with the current implementation. What do you think?